digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java

```diff
@@ -1328,6 +1465,14 @@ public void insertMonitors(long attemptId, List<SessionMonitor> monitors)
                 " values (:projectId, :workflowName, :sessionTime)")
         void upsertAndLockSession(@Bind("projectId") int projectId,
                 @Bind("workflowName") String workflowName, @Bind("sessionTime") long sessionTime);
+
+        @SqlQuery("select id from sessions s" +
+                " where id = (" +
+                    "select session_id from session_attempts" +
+                    " where id = :attemptId" +
+                ")" +
+                " for update")
```

Comment:
H2 can't use JOIN with FOR UPDATE. So, H2Dao.upsertAndLockSession returns only `id`.
PostgreSQL can use JOIN and can return more columns together (sa.* and s.*).